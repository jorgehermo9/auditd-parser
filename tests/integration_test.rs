use std::{fs, path::PathBuf};

use auditd_parser::{AuditdRecord, ParserError};
use erased_serde::Serialize;
use rstest::rstest;

#[rstest]
fn test_log_data(#[files("tests/data/**/*.log")] log_file: PathBuf) {
    let log_file_path = log_file.to_string_lossy().into_owned();
    let log_file_content = fs::read_to_string(log_file).unwrap();
    let logs = log_file_content
        .lines()
        .enumerate()
        .map(|(index, line)| (index + 1, line))
        .filter(|(_line_number, line)| !line.is_empty() && !line.starts_with('#'))
        .collect::<Vec<_>>();

    for (line_number, log) in logs {
        let maybe_record = log.parse::<AuditdRecord>();
        let location = format!("{log_file_path}:{line_number}");
        let log_identifier = get_log_identifier(log, &location);

        // Unwrap the Ok variant but leave the Err variant as-is
        let result: Box<dyn Serialize> = match maybe_record {
            Ok(record) => Box::new(record),
            Err(error) => Box::new(Err::<AuditdRecord, ParserError>(error)),
        };

        insta::with_settings!(
            {
                info => &log,
                description => location,
                snapshot_suffix => log_identifier,
            },
            {
                insta::assert_json_snapshot!(result);
            }
        );
    }
}

fn get_log_identifier(log: &str, location: &str) -> String {
    let log_md5 = md5::compute(format!("{log}{location}"));
    format!("{log_md5:x}")
}
