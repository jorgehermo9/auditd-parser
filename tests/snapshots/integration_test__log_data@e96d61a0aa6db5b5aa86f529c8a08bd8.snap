---
source: tests/integration_test.rs
expression: record
info: "type=SOCKADDR msg=audit(1490816924.990:517647): saddr=01002F7661722F72756E2F6E7363642F736F636B65740000603B7B47FC7F0000303C7B47FC7F0000020000000000000014000000160001030800000000000000C03B7B47FC7F0000103B7B47FC7F00002000000000000000303C7B47FC7F0000C0FB39861C7F0000787F2F861C7F"
---
{
  "record_type": "SOCKADDR",
  "timestamp": 1490816924990,
  "id": 517647,
  "fields": {
    "saddr": {
      "family": "AF_UNIX",
      "path": "/var/run/nscd/socket"
    }
  }
}
